_studio/shared/src/mfx_umc_mjpeg_vpp.cpp

```diff
@@ -169,77 +122,79 @@ UMC::Status mfx_UMC_FrameAllocator_D3D_Converter::InitMfx(UMC::FrameAllocatorPar
     return UMC::UMC_OK;
 }
 
+UMC::Status mfx_UMC_FrameAllocator_D3D_Converter::Reset()
+{
+    m_pCc.reset();
+    return mfx_UMC_FrameAllocator_D3D::Reset();
+}
+
 mfxStatus mfx_UMC_FrameAllocator_D3D_Converter::StartPreparingToOutput(mfxFrameSurface1 *surface_work,
                                                                        UMC::FrameData* in,
-                                                                       const mfxVideoParam * par,
-                                                                       VideoVppJpegD3D9 **pCc,
+                                                                       const mfxVideoParam *par,
                                                                        mfxU16 *taskId,
                                                                        bool isOpaq)
 {
     UMC::AutomaticUMCMutex guard(m_guard);
 
+    if(!m_pCc)
+    {
+        bool isD3DToSys = false;
+
+        if(par->IOPattern & MFX_IOPATTERN_OUT_SYSTEM_MEMORY)
+        {
+            isD3DToSys = true;
+        }
+        else if (par->IOPattern & MFX_IOPATTERN_OUT_OPAQUE_MEMORY)
+        {
+            mfxExtOpaqueSurfaceAlloc *pOpaqAlloc = (mfxExtOpaqueSurfaceAlloc *)GetExtendedBuffer(par->ExtParam, par->NumExtParam, MFX_EXTBUFF_OPAQUE_SURFACE_ALLOCATION);
+            MFX_CHECK(pOpaqAlloc, MFX_ERR_INVALID_VIDEO_PARAM);
+
+            isD3DToSys = (pOpaqAlloc->Out.Type & MFX_MEMTYPE_SYSTEM_MEMORY) != 0;
+        }
+
+        m_pCc.reset(new VideoVppJpegD3D9(m_pCore, isD3DToSys, par->IOPattern & MFX_IOPATTERN_OUT_OPAQUE_MEMORY));
+
+        mfxStatus mfxSts = m_pCc->Init(par);
+        MFX_CHECK_STS( mfxSts );
+    }
+
     mfxStatus sts = MFX_ERR_NONE;
-    bool isD3DToSys = false;
+
+    mfxHDLPair pPair;
+    if(isOpaq)
+        sts = m_pCore->GetFrameHDL(surface_work->Data.MemId, (mfxHDL*)&pPair);
+    else
+        sts = m_pCore->GetExternalFrameHDL(surface_work->Data.MemId, (mfxHDL*)&pPair);
 
     if(par->mfx.FrameInfo.PicStruct == MFX_PICSTRUCT_PROGRESSIVE)
     {
         UMC::FrameMemID index = in->GetFrameMID();
         mfxMemId memInter = m_frameDataInternal.GetSurface(index).Data.MemId;
         mfxMemId memId = isOpaq?(memInter):(m_pCore->MapIdx(memInter));
 
-        mfxHDLPair pPair;
-        if(isOpaq)
-            sts = m_pCore->GetFrameHDL(surface_work->Data.MemId, (mfxHDL*)&pPair);
-        else
-            sts = m_pCore->GetExternalFrameHDL(surface_work->Data.MemId, (mfxHDL*)&pPair);
+        // if memid of in is same as memid of surface_work, this method must not be called
+        MFX_CHECK_WITH_ASSERT(!pPair.first || pPair.first != memId, MFX_ERR_UNSUPPORTED);
+
+        mfxFrameSurface1 pSrc = m_frameDataInternal.GetSurface(index);
```

Comment:
Why `pSrc` if it's not actually a pointer?

Reply:
Done.